source/language/openpulse.rst

```diff
@@ -290,28 +298,262 @@ pushing the results into some buffer which is then accessed outside the program.
+      {...}
+  }
+
+Geometric gate
+~~~~~~~~~~~~~~
+
+.. code-block:: javascript
+
+  defcal prelude {
+      float[32] fq_01 = 5e9; // hardcode or pull from some function
+      float[32] anharm = 300e6; // hardcode or pull from some function
+      frame frame_01 = Frame(fq_01, 0);
+      frame frame_12 = Frame(fq_01 + anharm, 0);
+  }
+
+  defcal geo_gate(angle[32]: theta) %0 {
+      // theta: rotation angle (about z-axis) on Bloch sphere
+
+      tx_channel dq = txch($q, “drive”);
+
+      // Assume we have calibrated 0->1 pi pulses and 1->2 pi pulse
+      // envelopes (no sideband)
+      waveform X_01 = {...};
+      waveform X_12 = {...};
+      float[32] a = sin(theta/2);
+      float[32] b = sqrt(1-a**2);
+
+      // Double-tap
+      play(dq, scale(a, X_01), frame_01);
+      play(dq, scale(b, X_12), frame_12);
+      play(dq, scale(a, X_01), frame_01);
+      play(dq, scale(b, X_12), frame_12);
+  }
+
+Neutral atoms
+~~~~~~~~~~~~~
+
+In this simple example, the signal chain is composed of two electro-optic modulators (EOM) and
+an acousto-optic deflector (AOD). The EOMs put sidebands on the laser light while the AOD diffracts
+the light in an amount proportional to the frequency of the RF drive. This example was chosen
+because it is similar in spirit to the work by Levine et al. except that phase control is exerted
+using virtual Z gates on the AODs -- requiring frame tracking of the qubit frequency yet
+application of a tone that maps to the qubit position (i.e. requires the use of a sideband).
+
+The program aims to perform a Hahn echo sequence on q1, and a Ramsey sequence on q2 and q3.
+
+.. code-block:: javascript
+
+  defcal neutral_atoms {
+    // Define the channels
+    eom_a_channel = txch(0, "eom_a");
+    eom_a_channel = txch(1, "eom_b");
+    aod_channel = txch(0, "aod");
+
+    // Define the Raman frames, which are detuned by an amount Δ from the  5S1/2 to 5P1/2 transition
+    // and offset from each other by the qubit_freq
+    raman_a_frame = Frame(Δ, 0.0)
+    raman_b_frame = Frame(Δ-qubit_freq, 0.0)
+
+    // Waveforms supplied to the Raman beams are just constant
+
+
+    // Three copies of qubit freq to track phase of each qubit
+    q1_frame = Frame(qubit_freq, 0)
+    q2_frame = Frame(qubit_freq, 0)
+    q3_frame = Frame(qubit_freq, 0)
+
+    // Generic gaussian envelope
+    waveform π_half_sig = gaussian(..., π_half_time, ...)
+
+    // Waveforms ultimately supplied to the AODs. We mix our general Gaussian pulse with a sine wave to
+    // put a sideband on the signal construction to target the qubit position while maintainig the
+    // desired Rabi rate.
+    q1_π_half_sig = mix(π_half_sig, sine(q1_π_half_amp, q1_pos_freq-qubit_freq, 0.0, π_half_time));
+    q2_π_half_sig = mix(π_half_sig, sine(q2_π_half_amp, q2_pos_freq-qubit_freq, 0.0, π_half_time));
+    q3_π_half_sig = mix(π_half_sig, sine(q3_π_half_amp, q3_pos_freq-qubit_freq, 0.0, π_half_time));
+
+    for τ in [0: T]:
```

Comment:
Do we want loops within the `defcal`? You could make the `defcal` the pulse sequence within the loop

Reply:
Good question. I would say yes so long as they "have a definite length". Otherwise, I'm wondering where the setup information for the `q#_π_half_sig` would go in this example.  Do we allow for nested `defcal` blocks?

Reply:
yeah that's a fair point. It could be made global, but that might be too much. I think this is fine